src/metabase/driver/sql_jdbc/actions.clj

```diff
@@ -303,3 +303,57 @@
             [(conj errors {:index row-index, :error (ex-message e)})
              created-rows]))))
      rows)))
+
+(defn- pk-values->filters [table-id pk-values]
+  ;; check that all pk-values have the same shape
+  (when-not (apply = (map (comp set keys) pk-values))
+    (throw (ex-info (trs "Inconsistent row selection.")
+                    {:status-code 400, :pk-values pk-values})))
+  (let [given-pks (set (keys (first pk-values)))
+        table-pks (into {} (map (juxt :name :id) (filter #(= (:semantic_type %) :type/PK) (:fields (first (table/with-fields [(Table table-id)]))))))]
```

Comment:
When comparing `metabase.types` types you should use `isa?` instead of `=`. While we don't have any types that derive from `:type/PK` at this point in time, it's not impossible that we won't in the future

```suggestion
        table-pks (into {} (map (juxt :name :id) (filter #(isa? (:semantic_type %) :type/PK) (:fields (first (table/with-fields [(Table table-id)]))))))]
```